components/raftstore/src/store/snap.rs

```diff
@@ -1994,9 +1995,49 @@ impl TabletSnapManager {
         Ok(Self {
             base: path,
             receiving: Arc::default(),
+            stats: Arc::default(),
         })
     }
 
+    pub fn begin_snapshot(&self, key: TabletSnapKey, start: Instant, generate_duration_sec: u64) {
+        let mut stat = SnapshotStat::default();
+        stat.set_generate_duration_sec(generate_duration_sec);
+        self.stats.lock().unwrap().insert(key, (start, stat));
+    }
+
+    pub fn finish_snapshot(&self, key: TabletSnapKey, send: Instant) {
+        let region_id = key.region_id;
+        self.stats
+            .lock()
+            .unwrap()
+            .entry(key)
+            .and_modify(|(start, stat)| {
+                stat.set_send_duration_sec(start.saturating_elapsed().as_secs());
+                stat.set_total_duration_sec(send.saturating_elapsed().as_secs());
+                stat.set_region_id(region_id);
+            });
+    }
+
+    pub fn stats(&self) -> SnapStats {
+        let finished: HashMap<TabletSnapKey, (Instant, SnapshotStat)> = self
+            .stats
+            .lock()
+            .unwrap()
+            .drain_filter(|_, (_, stat)| stat.get_region_id() > 0)
+            .collect();
```

Comment:
The `collect` may be not necessary. `.drain_filter` return an iterator, and can be chained to following `map` and `filter`.

Reply:
fixed